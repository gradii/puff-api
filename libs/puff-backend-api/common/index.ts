export * from './src/project-repo.service'