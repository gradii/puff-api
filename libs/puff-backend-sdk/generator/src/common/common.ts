/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */

import * as fs from 'fs';
import * as fse from 'fs-extra';
import * as git from 'isomorphic-git';
import * as path from 'path';

//
export async function cloneFsRepo(
  sourceRepo: string,
  outputRepo: string,
  checkoutConfig = {}
) {
  const sourceGit = path.join(sourceRepo, '.git');
  const outPutGit = path.join(outputRepo, '.git');
  await fse.copy(sourceGit, outPutGit);

  await git.checkout({
    force: true,
    ref  : 'master',
    ...checkoutConfig,
    dir: outputRepo,
    fs
  });

}
