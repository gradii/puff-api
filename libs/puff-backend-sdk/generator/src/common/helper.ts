export * from './generate-helper/generate-import-declaration';
