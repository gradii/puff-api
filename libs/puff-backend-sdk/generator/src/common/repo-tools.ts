import * as fs from 'fs';
import * as json5 from 'json5';
import * as path from 'path';
import { intersection, uniq } from 'ramda';

export class RepoTools {
  public nxJsonInfo: any;
  public angularJsonInfo: any;
  private backendProjectName: any;
  private backendProjectInfo: any;

  constructor(private repo: string) {
    if (!fs.existsSync(path.join(repo, 'nx.json'))) {
      throw new Error(`nx.json not found in repo "${repo}"`);
    }
    if (!fs.existsSync(path.join(repo, 'angular.json'))) {
      throw new Error(`angular.json not found in repo "${repo}"`);
    }
    this.initRepoInfo();
  }

  initRepoInfo() {
    this.nxJsonInfo = json5.parse(fs.readFileSync(path.join(this.repo, 'nx.json'), { encoding: 'utf8' }));
    this.angularJsonInfo = json5.parse(fs.readFileSync(path.join(this.repo, 'angular.json'), { encoding: 'utf8' }));
  }

  private findProjectNames({
                             tags = [],
                             guessNames = [],
                             builder
                           }) {
    // const nxProjects = Object.keys(this.nxJsonInfo.projects)
    const names = [];
    const mayBeNames = [];
    for (let [key, val] of Object.entries<any>(this.nxJsonInfo.projects)) {
      if (val.tags && Array.isArray(val.tags) && intersection(val.tags, tags).length > 0) {
        names.push(key);
      }
    }
    for (let [key, val] of Object.entries<any>(this.angularJsonInfo.projects)) {
      if (val.projectType === 'application') {
        const buildInfo = val?.architect?.build;
        if (buildInfo?.builder === builder) {
          names.push(key);
        }
        if (guessNames.includes(key)) {
          mayBeNames.push(key);
        }
      }
    }

    return uniq([...names, ...mayBeNames]);
  }

  getProjectInfo(name) {
    return this.angularJsonInfo.projects[name];
  }

  setBackendProject(backendName?) {
    if (!backendName) {
      backendName = this.findProjectNames({ tags: ['backend'], builder: '@nrwl/node:build', guessNames: ['api', 'backend'] }).shift();
    }
    if (!backendName) {
      throw new Error(`can't find a backend project, specify a tag "backend" in nx.json file with project tags field`);
    }
    this.backendProjectName = backendName;
    this.backendProjectInfo = this.getProjectInfo(backendName);
    return this.backendProjectInfo;
  }

  getBackendProject(name?) {
    if (this.backendProjectInfo) {
      return this.backendProjectInfo;
    } else {
      return this.setBackendProject(name);
    }
  }

  getBackendProjectMainFile() {
    if (!this.backendProjectInfo) {
      this.setBackendProject();
    }
    return this.backendProjectInfo.architect.build.options.main;
  }

  getBackendProjectSourceRoot() {
    if (!this.backendProjectInfo) {
      this.setBackendProject();
    }
    return this.backendProjectInfo.sourceRoot;
  }


}
