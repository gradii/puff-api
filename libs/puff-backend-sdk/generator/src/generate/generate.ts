/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */

import * as fs from 'fs-extra';
import { join as pathJoin } from 'path';
import * as prettier from 'prettier';
import { Options } from 'prettier';
import * as ts from 'typescript';

/**
 *
 */
export abstract class Generate {
  imports = [];

  addedImports = [];

  abstract filename;
  abstract ctx;
  private prettierConfig: Options = {
    parser: 'typescript'
  };


  writeFile(file, content) {
    fs.outputFileSync(pathJoin(this.ctx.rootDir, file), content, { encoding: 'utf8' });
  }

  printFile(asts) {
    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed, omitTrailingSemicolon: true, removeComments: false });

    const resultFile = ts.createSourceFile(
      this.filename,
      '',
      ts.ScriptTarget.Latest,
      /*setParentNodes*/ false,
      ts.ScriptKind.TS);

    // @ts-ignore
    resultFile.statements = ts.createNodeArray(asts);
    const result = printer.printNode(ts.EmitHint.Unspecified, resultFile, resultFile);
    return result;
  }

  outputFile(asts, prettierConfig = {}) {
    const file = this.printFile(asts);
    const formatted = prettier.format(file, { ...this.prettierConfig, ...prettierConfig });
    this.writeFile(this.filename, formatted);
    // this.writeFile(this.filename, file);
  }

  protected _preGenerate() {

  }

  _normalizeImports() {
    // if (this.hasSwagger) {
    //   this.addedImports.push({
    //     from: '@nestjs/swagger', importClause: ['DocumentBuilder', 'SwaggerModule']
    //   });
    // }

    //region arrange imports order and combine import clause
    //endregion

    const rst: { [key: string]: Set<string> } = {};

    const namseSpaceImport = [];

    for (const it of [...this.imports, ...this.addedImports]) {
      // rst[it.from] = it.importClause;
      if (!rst[it.from]) {
        rst[it.from] = new Set();
      }
      if (Array.isArray(it.importClause)) {
        it.importClause.forEach(importClause => {
          rst[it.from].add(importClause);
        });
      } else if (typeof it.importClause === 'string') {
        namseSpaceImport.push(
          { from: it.from, importClause: it.importClause }
        );
      }
    }
    const rstList = [];
    for (const [key, val] of Object.entries(rst)) {
      rstList.push({ from: key, importClause: Array.from(val) });
    }

    return [...rstList.filter(it => it.importClause.length > 0), ...namseSpaceImport];
  }

  generate() {
    this._preGenerate();
    this._postGenerate();
  }

  protected _postGenerate() {
    this.addedImports = [];
  }
}
