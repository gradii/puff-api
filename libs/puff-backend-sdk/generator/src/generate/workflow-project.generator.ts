import { GenerateKillActionCommand } from './workflow/generate-kill-action-command';
import { GenerateKillActionCommandHandlerService } from './workflow/generate-kill-action-command-handler-service';
import { GenerateStepStatus } from './workflow/generate-step-status';
import { GenerateWorkflowModule } from './workflow/generate-workflow-module';
import { GenerateWorkflowSagaService } from './workflow/generate-workflow-saga-service';
import { GenerateWorkflowBaseEvent } from './workflow/generate-workflow-base-event';


export class WorkflowProjectGenerator {
  constructor(public ctx: {
    rootDir: string,
    projectName: string,
    workflowModel: any[]
  }) {
  }

  beginGenerate() {
    this.generateKillActionCommand();
    this.generateKillActionCommandHandlerService();
    this.generateStepStatus();
    this.generateWorkflowModule();
    this.generateWorkflowSagaService();
    this.generateWorkflowBaseEvent();

    //generate project

  }

  generateKillActionCommand() {
    const generate = new GenerateKillActionCommand(this.ctx);
    const asts = generate.generate();

    generate.outputFile(asts);
  }

  generateKillActionCommandHandlerService() {
    const generate = new GenerateKillActionCommandHandlerService(this.ctx);
    const asts = generate.generate();

    generate.outputFile(asts);
  }

  generateStepStatus() {
    const generate = new GenerateStepStatus(this.ctx);
    const asts = generate.generate();

    generate.outputFile(asts);
  }

  generateWorkflowModule() {
    const generate = new GenerateWorkflowModule(this.ctx);
    const asts = generate.generate();

    generate.outputFile(asts);
  }

  generateWorkflowSagaService() {
    const generate = new GenerateWorkflowSagaService(this.ctx);
    const asts = generate.generate();

    generate.outputFile(asts);
  }

  generateWorkflowBaseEvent() {
    const generate = new GenerateWorkflowBaseEvent(this.ctx);
    const asts = generate.generate();

    generate.outputFile(asts);
  }

}
