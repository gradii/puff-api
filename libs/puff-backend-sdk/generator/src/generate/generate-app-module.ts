import { Generate } from './generate';
import * as ts from 'typescript';
import { generateImportDeclaration } from '../common/helper';


export class GenerateAppModule extends Generate {
  filename = 'app/app.module.ts';

  imports = [
    { from: '@nestjs/common', importClause: ['Module'] },
    { from: '@nestjs/schedule', importClause: ['ScheduleModule'] },
    { from: './workflow/workflow.module', importClause: ['WorkflowModule'] },
    { from: './app.controller', importClause: ['AppController'] }
  ];

  addedImports = [];

  constructor(public ctx = {
    dependencies: []
  }) {
    super();
  }

  // _normalizeImports() {
  //
  // }

  _dummyMainAst() {
    const asts = [
      ts.createClassDeclaration(
        [ts.createDecorator(ts.createCall(
          ts.createIdentifier('Module'),
          undefined,
          [ts.createObjectLiteral(
            [
              ts.createPropertyAssignment(
                ts.createIdentifier('imports'),
                ts.createArrayLiteral(
                  [
                    ts.createCall(
                      ts.createPropertyAccess(
                        ts.createIdentifier('ScheduleModule'),
                        ts.createIdentifier('forRoot')
                      ),
                      undefined,
                      []
                    ),
                    ts.createIdentifier('WorkflowModule')
                  ],
                  true
                )
              ),
              ts.createPropertyAssignment(
                ts.createIdentifier('controllers'),
                ts.createArrayLiteral(
                  [ts.createIdentifier('AppController')],
                  false
                )
              ),
              ts.createPropertyAssignment(
                ts.createIdentifier('providers'),
                ts.createArrayLiteral(
                  [],
                  false
                )
              )
            ],
            true
          )]
        ))],
        [ts.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.createIdentifier('AppModule'),
        undefined,
        undefined,
        []
      )
    ];

    return [
      ...this._normalizeImports().map(it => {
        return generateImportDeclaration(it.from, it.importClause);
      }),
      ...asts
    ];
  }

  generate() {
    this._normalizeImports();
    this.ctx.dependencies = this.imports.map(it => it.from);

    const result = this._dummyMainAst();
    this._postGenerate();
    return result;
  }

  protected _postGenerate() {
    this.addedImports = [];
  }
}
