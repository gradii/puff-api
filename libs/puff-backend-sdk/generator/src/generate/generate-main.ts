import * as ts from 'typescript';
import { generateImportDeclaration } from '../common/helper';
import { Generate } from './generate';

export class GenerateMain extends Generate {

  filename = 'main.ts'

  private _hasHttp = true;
  get hasHttp(): boolean {
    return this._hasHttp;
  }

  set hasHttp(value: boolean) {
    this._hasHttp = value;
  }

  private _hasSwagger = false;
  get hasSwagger() {
    return this._hasHttp && this._hasSwagger;
  }

  set hasSwagger(value: boolean) {
    this._hasSwagger = value;
  }

  imports = [
    { from: '@nestjs/common', importClause: ['Logger'] },
    { from: '@nestjs/core', importClause: ['NestFactory'] },
    { from: './app/app.module', importClause: ['AppModule'] }
  ];

  addedImports = [];

  constructor(public ctx = {
    dependencies: []
  }) {
    super()
  }

  _normalizeImports() {
    if (this.hasSwagger) {
      this.addedImports.push({
        from: '@nestjs/swagger', importClause: ['DocumentBuilder', 'SwaggerModule']
      });
    }

    return []

    //region arrange imports order and combine import clause
    //endregion
  }

  _dummyMainAst() {
    return [
      //imports
      ...[...this.imports, ...this.addedImports].map(it => {
        return generateImportDeclaration(it.from, it.importClause);
      }),

      this._generateBootstrapFn(),

      //bootstrap()
      ts.createExpressionStatement(ts.createCall(
        ts.createIdentifier('bootstrap'),
        undefined,
        []
      ))
    ];
  }

  _generateSwaggerFn() {
    if (!this.hasSwagger) return [];

    return [
      ts.addSyntheticLeadingComment(
        ts.createNotEmittedStatement(ts.createEmptyStatement()),
        ts.SyntaxKind.SingleLineCommentTrivia, 'region swagger api'),
      ts.createVariableStatement(
        undefined,
        ts.createVariableDeclarationList(
          [ts.createVariableDeclaration(
            ts.createIdentifier('apiPath'),
            undefined,
            ts.createStringLiteral('/swagger-api')
          )],
          ts.NodeFlags.Const | ts.NodeFlags.AwaitContext | ts.NodeFlags.ContextFlags | ts.NodeFlags.TypeExcludesFlags
        )
      ),
      ts.createVariableStatement(
        undefined,
        ts.createVariableDeclarationList(
          [ts.createVariableDeclaration(
            ts.createIdentifier('options'),
            undefined,
            ts.createCall(
              ts.createPropertyAccess(
                ts.createCall(
                  ts.createPropertyAccess(
                    ts.createCall(
                      ts.createPropertyAccess(
                        ts.createCall(
                          ts.createPropertyAccess(
                            ts.createCall(
                              ts.createPropertyAccess(
                                ts.createNew(
                                  ts.createIdentifier('DocumentBuilder'),
                                  undefined,
                                  []
                                ),
                                ts.createIdentifier('setTitle')
                              ),
                              undefined,
                              [ts.createStringLiteral('api')]
                            ),
                            ts.createIdentifier('setDescription')
                          ),
                          undefined,
                          [ts.createStringLiteral('API description ')]
                        ),
                        ts.createIdentifier('setVersion')
                      ),
                      undefined,
                      [ts.createStringLiteral('1.0')]
                    ),
                    ts.createIdentifier('addBearerAuth')
                  ),
                  undefined,
                  []
                ),
                ts.createIdentifier('build')
              ),
              undefined,
              []
            )
          )],
          ts.NodeFlags.Const | ts.NodeFlags.AwaitContext | ts.NodeFlags.ContextFlags | ts.NodeFlags.TypeExcludesFlags
        )
      ),
      ts.createVariableStatement(
        undefined,
        ts.createVariableDeclarationList(
          [ts.createVariableDeclaration(
            ts.createIdentifier('document'),
            undefined,
            ts.createCall(
              ts.createPropertyAccess(
                ts.createIdentifier('SwaggerModule'),
                ts.createIdentifier('createDocument')
              ),
              undefined,
              [
                ts.createIdentifier('app'),
                ts.createIdentifier('options')
              ]
            )
          )],
          ts.NodeFlags.Const | ts.NodeFlags.AwaitContext | ts.NodeFlags.ContextFlags | ts.NodeFlags.TypeExcludesFlags
        )
      ),
      ts.createExpressionStatement(ts.createCall(
        ts.createPropertyAccess(
          ts.createIdentifier('SwaggerModule'),
          ts.createIdentifier('setup')
        ),
        undefined,
        [
          ts.createIdentifier('apiPath'),
          ts.createIdentifier('app'),
          ts.createIdentifier('document')
        ]
      )),
      ts.addSyntheticLeadingComment(
        ts.createNotEmittedStatement(ts.createEmptyStatement()),
        ts.SyntaxKind.SingleLineCommentTrivia, 'endregion')
    ];
  }

  _generateBootstrapFn() {
    return ts.createFunctionDeclaration(
      undefined,
      [ts.createModifier(ts.SyntaxKind.AsyncKeyword)],
      undefined,
      ts.createIdentifier('bootstrap'),
      undefined,
      [],
      undefined,
      ts.createBlock(
        [
          ...(/* function generateHasHttp */() => {
            if (!this.hasHttp) return [];
            return [
              ts.createVariableStatement(
                undefined,
                ts.createVariableDeclarationList(
                  [ts.createVariableDeclaration(
                    ts.createIdentifier('app'),
                    undefined,
                    ts.createAwait(ts.createCall(
                      ts.createPropertyAccess(
                        ts.createIdentifier('NestFactory'),
                        ts.createIdentifier('create')
                      ),
                      undefined,
                      [ts.createIdentifier('AppModule')]
                    ))
                  )],
                  ts.NodeFlags.Const | ts.NodeFlags.AwaitContext | ts.NodeFlags.ContextFlags | ts.NodeFlags.TypeExcludesFlags
                )
              ),
              ts.createVariableStatement(
                undefined,
                ts.createVariableDeclarationList(
                  [ts.createVariableDeclaration(
                    ts.createIdentifier('globalPrefix'),
                    undefined,
                    ts.createStringLiteral('api')
                  )],
                  ts.NodeFlags.Const | ts.NodeFlags.AwaitContext | ts.NodeFlags.ContextFlags | ts.NodeFlags.TypeExcludesFlags
                )
              ),
              ts.createExpressionStatement(ts.createCall(
                ts.createPropertyAccess(
                  ts.createIdentifier('app'),
                  ts.createIdentifier('setGlobalPrefix')
                ),
                undefined,
                [ts.createIdentifier('globalPrefix')]
              )),
              ...this._generateSwaggerFn(),
              ts.createVariableStatement(
                undefined,
                ts.createVariableDeclarationList(
                  [ts.createVariableDeclaration(
                    ts.createIdentifier('port'),
                    undefined,
                    ts.createBinary(
                      ts.createPropertyAccess(
                        ts.createPropertyAccess(
                          ts.createIdentifier('process'),
                          ts.createIdentifier('env')
                        ),
                        ts.createIdentifier('PORT')
                      ),
                      ts.createToken(ts.SyntaxKind.BarBarToken),
                      ts.createNumericLiteral('3333')
                    )
                  )],
                  ts.NodeFlags.Const | ts.NodeFlags.AwaitContext | ts.NodeFlags.ContextFlags | ts.NodeFlags.TypeExcludesFlags
                )
              ),
              ts.createExpressionStatement(ts.createAwait(ts.createCall(
                ts.createPropertyAccess(
                  ts.createIdentifier('app'),
                  ts.createIdentifier('listen')
                ),
                undefined,
                [
                  ts.createIdentifier('port'),
                  ts.createArrowFunction(
                    undefined,
                    undefined,
                    [],
                    undefined,
                    ts.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
                    ts.createBlock(
                      [ts.createExpressionStatement(ts.createCall(
                        ts.createPropertyAccess(
                          ts.createIdentifier('Logger'),
                          ts.createIdentifier('log')
                        ),
                        undefined,
                        [ts.createBinary(
                          ts.createBinary(
                            ts.createBinary(
                              ts.createStringLiteral('Listening at http://localhost:'),
                              ts.createToken(ts.SyntaxKind.PlusToken),
                              ts.createIdentifier('port')
                            ),
                            ts.createToken(ts.SyntaxKind.PlusToken),
                            ts.createStringLiteral('/')
                          ),
                          ts.createToken(ts.SyntaxKind.PlusToken),
                          ts.createIdentifier('globalPrefix')
                        )]
                      ))],
                      true
                    )
                  )
                ]
              )))
            ];
          })(),
          ...(/*function generateNotHasHttp*/() => {
            if (this.hasHttp) return [];
            return [
              ts.createVariableStatement(
                undefined,
                ts.createVariableDeclarationList(
                  [ts.createVariableDeclaration(
                    ts.createIdentifier('app'),
                    undefined,
                    ts.createAwait(ts.createCall(
                      ts.createPropertyAccess(
                        ts.createIdentifier('NestFactory'),
                        ts.createIdentifier('createApplicationContext')
                      ),
                      undefined,
                      [ts.createIdentifier('AppModule')]
                    ))
                  )],
                  ts.NodeFlags.Const | ts.NodeFlags.AwaitContext | ts.NodeFlags.ContextFlags | ts.NodeFlags.TypeExcludesFlags
                )
              ),
              ...this._generateSwaggerFn(),
              ts.createExpressionStatement(ts.createCall(
                ts.createPropertyAccess(
                  ts.createIdentifier('Logger'),
                  ts.createIdentifier('log')
                ),
                undefined,
                [ts.createStringLiteral('Server running without http server')]
              ))
            ];
          })(),

          ts.createReturn(ts.createIdentifier('app'))
        ],
        true
      )
    );
  }

  generate() {
    this._normalizeImports();
    this.ctx.dependencies = this.imports.map(it => it.from);

    const result = this._dummyMainAst();
    this._postGenerate();
    return result;
  }

  protected _postGenerate() {
    this.addedImports = [];
  }
}
