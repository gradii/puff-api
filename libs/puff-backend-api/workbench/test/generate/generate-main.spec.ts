import { buildConfig } from 'package-tools';
import * as path from 'path';
import * as fse from 'fs-extra';
import { cloneFsRepo } from '../../lib/common/common';
import { generateRepoPath } from '../common';
import * as ts from 'typescript';
import { GenerateMain } from '../../lib/generate/generate-main';
import * as prettier from 'prettier';
import * as fs from 'fs';

const root = buildConfig.projectDir;
const dummyRepo = path.join(root, 'storage/_workbench/dummy-repo');

describe('backend bootstrap', () => {
  const testProjectRepo = generateRepoPath();

  beforeAll(async () => {
    await fse.ensureDirSync(testProjectRepo);
    await cloneFsRepo(dummyRepo, testProjectRepo);
  });

  afterAll(async () => {
    await fse.remove(testProjectRepo);
  });

  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed, omitTrailingSemicolon: true,  removeComments: false });

  let resultFile: ts.SourceFile;
  beforeEach(() => {
    resultFile = ts.createSourceFile('someFileName.ts', '', ts.ScriptTarget.Latest, /*setParentNodes*/ false, ts.ScriptKind.TS);
  });

  it('test generate main.ts', () => {
    const generateMain = new GenerateMain();

    resultFile.statements = ts.createNodeArray(generateMain.generate());

    const result = printer.printNode(ts.EmitHint.Unspecified, resultFile, resultFile);

    expect(result).toMatchSnapshot();

    const prettierConfig = JSON.parse(fs.readFileSync(path.join(testProjectRepo, '.prettierrc'), { encoding: 'utf8' }));

    const formatted = prettier.format(result, { ...prettierConfig, parser: 'typescript' });

    expect(formatted).toMatchSnapshot();
  });

  it('test generate main.ts with hasHttp false', () => {
    const generateMain = new GenerateMain();
    generateMain.hasHttp = false;

    resultFile.statements = ts.createNodeArray(generateMain.generate());

    const result = printer.printNode(ts.EmitHint.Unspecified, resultFile, resultFile);

    expect(result).toMatchSnapshot();

    const prettierConfig = JSON.parse(fs.readFileSync(path.join(testProjectRepo, '.prettierrc'), { encoding: 'utf8' }));

    const formatted = prettier.format(result, { ...prettierConfig, parser: 'typescript' });

    expect(formatted).toMatchSnapshot();
  });

  it('test generate main.ts with swagger', () => {
    const generateMain = new GenerateMain();
    generateMain.hasHttp = true;
    generateMain.hasSwagger = true;

    resultFile.statements = ts.createNodeArray(generateMain.generate());

    const result = printer.printNode(ts.EmitHint.Unspecified, resultFile, resultFile);

    expect(result).toMatchSnapshot('hasHttp = true; hasSwagger = true;');

    const prettierConfig = JSON.parse(fs.readFileSync(path.join(testProjectRepo, '.prettierrc'), { encoding: 'utf8' }));

    const formatted = prettier.format(result, { ...prettierConfig, parser: 'typescript' });

    expect(formatted).toMatchSnapshot('hasHttp = true; hasSwagger = true; prettier!');

    generateMain.hasHttp = false;

    resultFile.statements = ts.createNodeArray(generateMain.generate());
    const resultWithoutHttp = printer.printNode(ts.EmitHint.Unspecified, resultFile, resultFile);

    expect(resultWithoutHttp).toMatchSnapshot('hasHttp false; hasSwagger = true;');
  });

});
