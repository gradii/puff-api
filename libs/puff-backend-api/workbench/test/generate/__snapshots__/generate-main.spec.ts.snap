// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backend bootstrap test generate main.ts 1`] = `
"import { Logger } from \\"@nestjs/common\\";
import { NestFactory } from \\"@nestjs/core\\";
import { AppModule } from \\"./app/app.module\\";
async function bootstrap() {
    const app = await NestFactory.create(AppModule);
    const globalPrefix = \\"api\\";
    app.setGlobalPrefix(globalPrefix);
    const port = process.env.PORT || 3333;
    await app.listen(port, () => {
        Logger.log(\\"Listening at http://localhost:\\" + port + \\"/\\" + globalPrefix);
    });
    return app;
}
bootstrap();
"
`;

exports[`backend bootstrap test generate main.ts 2`] = `
"import { Logger } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app/app.module';
async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  const globalPrefix = 'api';
  app.setGlobalPrefix(globalPrefix);
  const port = process.env.PORT || 3333;
  await app.listen(port, () => {
    Logger.log('Listening at http://localhost:' + port + '/' + globalPrefix);
  });
  return app;
}
bootstrap();
"
`;

exports[`backend bootstrap test generate main.ts with hasHttp false 1`] = `
"import { Logger } from \\"@nestjs/common\\";
import { NestFactory } from \\"@nestjs/core\\";
import { AppModule } from \\"./app/app.module\\";
async function bootstrap() {
    const app = await NestFactory.createApplicationContext(AppModule);
    Logger.log(\\"Server running without http server\\");
    return app;
}
bootstrap();
"
`;

exports[`backend bootstrap test generate main.ts with hasHttp false 2`] = `
"import { Logger } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app/app.module';
async function bootstrap() {
  const app = await NestFactory.createApplicationContext(AppModule);
  Logger.log('Server running without http server');
  return app;
}
bootstrap();
"
`;

exports[`backend bootstrap test generate main.ts with swagger: hasHttp = true; hasSwagger = true; 1`] = `
"import { Logger } from \\"@nestjs/common\\";
import { NestFactory } from \\"@nestjs/core\\";
import { AppModule } from \\"./app/app.module\\";
import { DocumentBuilder, SwaggerModule } from \\"@nestjs/swagger\\";
async function bootstrap() {
    const app = await NestFactory.create(AppModule);
    const globalPrefix = \\"api\\";
    app.setGlobalPrefix(globalPrefix);
    //region swagger api
    const apiPath = \\"/swagger-api\\";
    const options = new DocumentBuilder().setTitle(\\"api\\").setDescription(\\"API description \\").setVersion(\\"1.0\\").addBearerAuth().build();
    const document = SwaggerModule.createDocument(app, options);
    SwaggerModule.setup(apiPath, app, document);
    //endregion
    const port = process.env.PORT || 3333;
    await app.listen(port, () => {
        Logger.log(\\"Listening at http://localhost:\\" + port + \\"/\\" + globalPrefix);
    });
    return app;
}
bootstrap();
"
`;

exports[`backend bootstrap test generate main.ts with swagger: hasHttp = true; hasSwagger = true; prettier! 1`] = `
"import { Logger } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app/app.module';
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';
async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  const globalPrefix = 'api';
  app.setGlobalPrefix(globalPrefix);
  //region swagger api
  const apiPath = '/swagger-api';
  const options = new DocumentBuilder()
    .setTitle('api')
    .setDescription('API description ')
    .setVersion('1.0')
    .addBearerAuth()
    .build();
  const document = SwaggerModule.createDocument(app, options);
  SwaggerModule.setup(apiPath, app, document);
  //endregion
  const port = process.env.PORT || 3333;
  await app.listen(port, () => {
    Logger.log('Listening at http://localhost:' + port + '/' + globalPrefix);
  });
  return app;
}
bootstrap();
"
`;

exports[`backend bootstrap test generate main.ts with swagger: hasHttp false; hasSwagger = true; 1`] = `
"import { Logger } from \\"@nestjs/common\\";
import { NestFactory } from \\"@nestjs/core\\";
import { AppModule } from \\"./app/app.module\\";
async function bootstrap() {
    const app = await NestFactory.createApplicationContext(AppModule);
    Logger.log(\\"Server running without http server\\");
    return app;
}
bootstrap();
"
`;
